### ISSUE TITLE
Similar performance between mmapfs and nativeunixfs

### ISSUE THREAD

**Issue by micoq** [micoq/native-unix-store#1]
Thankyou @micoq 
I did a write test using esrally, this is my test environment, arm architecture, 4cpu, 16g memory, -Xms8g, -Xmx8g, the data set size I wrote is 3g,between each test,i restarted OS, test results:

  | mmapfs | niofs | nativeunixfs
-- | -- | -- | --
Median Throughput（docs/s） | 33449.35 | 33186.3 | 31830.4
I did a lot of tests and the results are the same.According to the results, the plugin performance is not effective.What is wrong here?

_Originally posted by @azuresky11 in https://github.com/elastic/elasticsearch/issues/27748#issuecomment-482551555_

**Comment by micoq** [comment]
Hello @azuresky11
There is probably nothing wrong. In your case the data set can completely be stored in the filesystem cache (let's say ~7 GB). Since the main purpose of the plugin is to save the cache, the performance will be similar in this case (or a little worse since the data will not be read ahead).

If not already done, you could also try to enable direct I/O.

Is it a public dataset you use in your tests ? (a track included in esrally ?)

The major difference should be on reading. The plugin is made to avoid unnecessary reads as seen in this post:
https://discuss.elastic.co/t/when-to-use-mmapfs-and-niofs/176536

**Comment by azuresky11** [comment]
@micoq
It is a public dataset.And "index.refresh_interval" : "1".Based on the above configuration，I did another test, there are some ways to make the memory not enough.test results:

  | mmapfs | niofs | nativeunixfs
-- | -- | -- | --
Median Throughput | 25437.7 | 23450.4 | 20603.7

How can I use JDK8 to compile the plugin?

**Comment by azuresky11** [comment]
@micoq
It is a public dataset.And "index.refresh_interval" : "1".Based on the above configuration，I did another test, there are some ways to make the memory not enough.test results:
  | mmapfs | niofs | nativeunixfs
-- | -- | -- | --
Median Throughput | 25437.7 | 23450.4 | 20603.7

I am very confused about this result.

Another question，How can I use JDK8 to compile the plugin?

**Comment by azuresky11** [comment]
@micoq
It is a public dataset.And "index.refresh_interval" : "1".Based on the above configuration，I did another test, there are some ways to make the memory not enough.test results:
  | mmapfs | niofs | nativeunixfs
-- | -- | -- | --
Median Throughput | 25437.7 | 21654.4 | 20603.7

I am very confused about this result.

Another question，How can I use JDK8 to compile the plugin?

**Comment by azuresky11** [comment]
@micoq
It is a public dataset.And "index.refresh_interval" : "1".Based on the above configuration，I did another test, there are some ways to make the memory not enough.test results:

  | mmapfs | niofs | nativeunixfs
-- | -- | -- | --
Median Throughput | 25437.7 | 23450.4 | 20603.7

I am very confused about this result.

Another question，How can I use JDK8 to compile the plugin?

**Comment by micoq** [comment]
If your dataset is provided by esrally, what is the name of the track you used ? (as given by `esrally list tracks`)

`"index.refresh_interval" : "1"` should raise an error, it's either:
- `"index.refresh_interval" : "-1"`: no refresh at all (except when the indexing buffer is full and must be flushed)
- `"index.refresh_interval" : "1s"`: a refresh of one second. This leads to a very poor throughput since this creates a lot of segments (see a benchmark here https://sematext.com/blog/elasticsearch-refresh-interval-vs-indexing-performance, it's an old ES version but it's still relevant).

If you used the value of `-1`, the results are very strange. Is the refresh interval the only index setting you  change between the two tests ?
About 30% of the throughput is lost in the second test regardless the store type:

|| mmapfs | niofs | nativeunixfs
| --- | --- | --- | ---
| Test 1 | 33449.35 | 33186.3 | 31830.4
| Test 2 | 25437.7 | 23450.4 | 20603.7

It could be interesting to collect the system stats while you run the benchmark to find the bottleneck. The indexing speed depend on many factors like the refresh interval, the indexing buffer size (https://www.elastic.co/guide/en/elasticsearch/reference/master/indexing-buffer.html) or the bulk/write pool size (https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-threadpool.html).

**Comment by azuresky11** [comment]
I did not change "index.refresh_interval",it is the default,the two tests are the same.The only difference is that test1 has enough memory and test2 has insufficient memory.

**Comment by micoq** [comment]
@azuresky11 
The compilation uses the build-tools provided by Elastic (through gradle) and seems to require a specific JDK to run (the minimum version depends on the version of Elasticsearch). However the binaries are compatible with Java 8 (`targetCompatibility = 1.8`)

For the tests, I ran several queries against a index containing logs from various software, essentially filters (`query_string`). I published the results in the original issue (https://github.com/elastic/elasticsearch/issues/27748).
As you can see, the results are not necessary better (for the query speed) but a lot of filesystem cache is saved.

For example, a request like `name:*s*` on `geonames`  will load approximately 600MB of data in the cache for only 200MB actually used:
```
{"size":0,"query":{"query_string":{"query":"name:*s*"}}}
```
This could lead to large sequential reads from the storage and finally high latencies on queries.

Unfortunately, it's difficult to reproduce this behavior with simple benchmarks. In my case, I use this plugin in a production environment where we have been facing issues like this and discovered the cache problem. Another good explanation here: https://phabricator.wikimedia.org/T169498

Basically, the plugin disables the readahead so the storage must do much more IOs per second to loads the same data. In some cases, it's beneficial and in some other cases, it's not.

In your case, if you do not have this problem, I recommend you to start with `mmapfs` and monitor you disk IO and throughput.

However, the results of your last test (aggregation) is strange, `niofs` should be slower than `mmapfs`.

**Comment by azuresky11** [comment]
@micoq 
I open es with jdk8,but there one error:
java.lang.UnsupportedClassVersionError: fr/micoq/unsafe/MappedMemory has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0.

The result of last test (aggregation)  are first test after opened es,it is cold data,"cold" is made after a fresh startup and empty caches (echo 3 > /proc/sys/vm/drop_caches).And warmup test result:niofs  slower than mmapfs.

I use esrally for query testing,Mmapfs, niofs, and nativeunixfs have almost the same performance.Very different from 1t dataset query test.

**Comment by micoq** [comment]
@azuresky11 
For the Java 8 incompatibility it was a bug. I fixed it (only the `unsafe-tools` package was incorrectly compiled for Java 11+).
I also backported the changes made on the last version (1.1.1). This should improve the indexing speed (should be similar to `mmapfs`).

You can try the newer version `v6.6.0-1.1.1`

**Comment by micoq** [comment]
@azuresky11 
You could also try to tune the readahead size of your Elasticsearch data partition:
- a readahead of 0 should give you the same performances as using the plugin (without directIO)
- a large readahead will consume more filesystem cache but could be more profitable in some cases

To set the readahead:
`blockdev --setra 256 /dev/sda1` for a readahead of 256 blocks (128KB)
(with LVM, you must change the logiciel volume readahead)

I made a comment about that in the issue opened on the Elasticsearch page:
https://github.com/elastic/elasticsearch/issues/27748#issuecomment-484885110

**Comment by azuresky11** [comment]
@micoq thank you so much!
I build the  newer version plugin,but the zip's version is 7.0.0-1.1.1,and i can not use this version on elasticsearch-6.6.0.there are something wrong here.

**Comment by micoq** [comment]
@azuresky11 
It's fixed: the version tag was not set to the right commit point. You can download the sources again for the release `v6.6.0-1.1.1`
